File: src/main/java/editor/gui/filter/editor/DefaultsFilterPanel.java
package editor.gui.filter.editor;

import java.awt.FlowLayout;
import java.util.HashMap;
import java.util.Map;

import scala.jdk.javaapi.CollectionConverters;

import editor.collection.deck.Category;
import editor.filter.Filter;
import editor.filter.leaf.FilterLeaf;
import editor.gui.generic.ComboBoxPanel;
import editor.gui.settings.SettingsDialog;

/**
 * This class represents a panel that presents the user with a
 * single combo box showing the default categories in the
 * settings dialog and returns the filter for the selected one.
 * It is replaced with the contents of the selected filter in the
 * filter tree.
 *
 * @author Alec Roelke
 */
public class DefaultsFilterPanel extends FilterEditorPanel<FilterLeaf<?>>
{
    /**
     * Combo box showing the default categories.
     */
    private ComboBoxPanel<String> defaults;
    /**
     * Map of category names onto their String representations.
     */
    private Map<String, Category> categories;

    /**
     * Create a new DefaultsFilterPanel.
     */
    public DefaultsFilterPanel()
    {
        super();
        setLayout(new FlowLayout(FlowLayout.LEFT, 0, 0));

        categories = new HashMap<>();

        var presets = SettingsDialog.settings().editor().categories().presets();
        String[] names = CollectionConverters.asJava(presets).stream().map(Category::getName).toArray(String[]::new);
        for (int i = 0; i < presets.size(); i++)
            categories.put(names[i], presets.apply(i));

        defaults = new ComboBoxPanel<>(names);
        add(defaults);
    }

    /**
     * {@inheritDoc}
     * This panel doesn't have its own filter; rather, it returns a group containing
     * a filter corresponding to the given preset.
     */
    @Override
    public Filter filter()
    {
        return new Category(categories.get(defaults.getSelectedItem())).getFilter();
    }

    /**
     * {@inheritDoc}
     * Since this panel doesn't correspond to a filter, this doesn't do anything.
     */
    @Override
    public void setContents(FilterLeaf<?> filter)
    {}
}

File: src/main/scala/editor/gui/filter/editor/DefaultsFilterPanel.scala
package editor.gui.filter.editor

import editor.collection.deck.Category
import editor.database.attributes.CardAttribute
import editor.filter.FilterGroup
import editor.filter.leaf.FilterLeaf
import editor.gui.generic.ComboBoxPanel
import editor.gui.settings.SettingsDialog

import java.awt.FlowLayout
import scala.collection.immutable.ListMap

/**
 * Panel that contains shortcuts for adding filters based on the filters of the preset categories. This panel
 * doesn't actually correspond to a filter; internally, it is replaced by the contents of the filter of the
 * selected preset category.  If the filter is viewed again, this panel will be replaced by a group containing
 * the selected preset category's filter.
 *
 * @author Alec Roelke
 */
class DefaultsFilterPanel extends FilterEditorPanel[FilterLeaf] {
  setLayout(FlowLayout(FlowLayout.LEFT, 0, 0))

  private val categories = ListMap.from(SettingsDialog.settings.editor.categories.presets.map((p) => p.getName -> p))
  private val defaults = ComboBoxPanel(categories.keys.toArray)
  add(defaults)

  override val attribute = CardAttribute.DEFAULTS
  override def filter = FilterGroup(Seq(Category(categories(defaults.getSelectedItem)).getFilter))
  override def setFields(filter: FilterLeaf) = throw UnsupportedOperationException("defaults filter panel should be replaced by contents")
}